scripts/core/editor3/components/MultipleHighlights.jsx

```diff
@@ -119,70 +119,132 @@ export class MultipleHighlights extends React.Component {
         return state;
     }
 
-    addHighlight(highlightType, highlightData) {
-        // NICE-TO-HAVE: remove associated data when highlights are removed by deleting characters in the editor
+    setHighlightsState(editorState, highlightsState) {
+        // TODO: save to first block
+        this.setState(highlightsState);
+    }
+
+    getHighlightsStates(editorState) {
+        // TODO: get from first block
+        return this.state;
+    }
+
+    saveHighlights({newEditorState, newHighlightsState}, applyChanges) {
+        this.setHighlightsState(newEditorState, newHighlightsState);
+
+        if (applyChanges) {
+            this.props.onHighlightChange(newEditorState, this.exportState());
+        }
+
+        return newEditorState;
+    }
+
+    initHighlights(editorState) {
+        const applyChanges = editorState === undefined;
+        const newEditorState = editorState || this.props.editorState;
+        const newHighlightsState = this.getHighlightsStates(editorState);
+
+        return {newEditorState, newHighlightsState, applyChanges};
+    }
 
-        const styleName = highlightType + '-' + (this.state.lastHighlightIds[highlightType] + 1);
+    _addHighlight(editorState, highlightsState, highlightType, highlightData) {
+        // NICE-TO-HAVE: remove associated data when highlights are removed by deleting characters in the editor
+        const styleName = highlightType + '-' + (highlightsState.lastHighlightIds[highlightType] + 1);
 
-        this.setState({
+        const newHighlightsState = {
             lastHighlightIds: {
-                ...this.state.lastHighlightIds,
-                [highlightType]: this.state.lastHighlightIds[highlightType] + 1
+                ...highlightsState.lastHighlightIds,
+                [highlightType]: highlightsState.lastHighlightIds[highlightType] + 1
             },
             highlightsStyleMap: {
-                ...this.state.highlightsStyleMap,
+                ...highlightsState.highlightsStyleMap,
                 [styleName]: this.props.availableHighlights[highlightType]
             },
-            highlightsData: {...this.state.highlightsData, [styleName]: highlightData}
-        }, () => {
-            this.props.onHighlightChange(
-                RichUtils.toggleInlineStyle(
-                    this.props.editorState,
-                    styleName
-                ),
-                this.exportState()
-            );
-        });
-    }
-
-    removeHighlight(styleName) {
+            highlightsData: {
+                ...highlightsState.highlightsData,
+                [styleName]: highlightData
+            }
+        };
+
+        const newEditorState = RichUtils.toggleInlineStyle(editorState, styleName);
+
+        return {newEditorState, newHighlightsState};
+    }
+
+
+    addHighlight(highlightType, highlightData, editorState) {
+        const {newEditorState, newHighlightsState, applyChanges} = this.initHighlights(editorState);
```

Comment:
I think it'd be better if you used `this.getHighlightsStates` here since it's all `initHighlights` does.

Reply:
The same code appear 4 times or more, so I created a function for it.

Reply:
If we send all the time the editorState, yes, we can drop initHighlights.